src/java/org/apache/cassandra/db/compaction/CompactionCursor.java

```
@@ -0,0 +1,1573 @@
+                sb.append(i + 1).append(":").append(histogram[i]).append(", ");
+                sum += (i + 1) * histogram[i];
+            }
+        }
+        if (sb.length() > 1)
+            sb.setLength(sb.length() - 1); //trim trailing comma
+        sb.append("] = " + sum);
+        return sb.toString();
+    }
+
+    private void retainPrevKeyForValidation()
+    {
+        // swap the reusable keys
+        if (prevKey == null)
+        {
+            prevKey = new ReusableDecoratedKey(new ReusableLongToken());
+        }
+        ReusableDecoratedKey temp = prevKey;
+        prevKey = partitionKey;
+        sstableCursors[0].currentDecoratedKey = temp;
+    }
+
+    public void close()
+    {
+        finish();
+        try
+        {
+            for (SSTableCursorReader reader : sstableCursors)
+            {
+                reader.close();
+            }
+        }
+        finally
+        {
+            activeCompactions.finishCompaction(this);
+        }
+
+        if (LOGGER.isInfoEnabled())
+        {
+            long position = ssTableCursorWriter == null ? 0 : ssTableCursorWriter.getPosition();
+            LOGGER.info("Compaction ended {}: { data bytes read = {}, data bytes written = {}, " +
+                        " input (keys = {}, rows = {}, cells = {}), " +
+                        " output (keys = {}, rows = {}, cells = {})}",
+                        this.compactionId, getTotalBytesScanned(), position + totalDataBytesWritten,
+                        mergeHistogramToString(partitionMergeCounters), mergeHistogramToString(rowMergeCounters), mergeHistogramToString(cellMergeCounters),
+                        sumHistogram(partitionMergeCounters), sumHistogram(rowMergeCounters), sumHistogram(cellMergeCounters));
+        }
+    }
+
+    public static <T> void bubbleInsertToPreSorted(T[] preSortedArray, boolean[] equalsNext, int sortedFrom, int sortedTo, Comparator<T> comparator){
+        if (sortedFrom == 0)
+            return;
```

I would throw an exception in this case; consider passing the index of the element we are bubbling down instead of the start of the sorted chain.